### PR TITLE
Add Nordic Uart Service (NUS) peripheral to examples

### PR DESCRIPTION
**Pull request by plaes**
Add generic peripheral implementation of Nordic Uart Service (NUS) and add example of it to nrf-sdc.

Has still some TODOs

### CODE REVIEW

File: examples/apps/src/ble_nus_peripheral.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    let mut table: AttributeTable<'_, NoopRawMutex, MAX_ATTRIBUTES> = AttributeTable::new();
+
+    // Generic Access Service (mandatory)
+    let id = b"Trouble Example Device";
+    let mut svc = table.add_service(Service::new(0x1800));
+    let _ = svc.add_characteristic_ro(0x2a00, id);
+    svc.build();
+
+    // Generic attribute service (mandatory)
+    table.add_service(Service::new(0x1801));
```
Comment (lulf) [inline examples/apps/src/ble_nus_peripheral.rs:62]: Just tagging you here @petekubiak and @jamessizeland, I wonder if it would make sense to have the macros generate these mandatory services? Iirc the old softdevice does that internally.

Comment (petekubiak) [reply examples/apps/src/ble_nus_peripheral.rs:62]: We discussed this when working on the last PR, and are definitely keen to implement something along these lines at some point.
I'm not an expert in BLE so excuse if this is a silly question, but my understanding is that these services are associated with GAP for a generic BLE device. Is this something we would always want included, or instead an optional feature if the user wants to implement GAP in their application?

Comment (lulf) [reply examples/apps/src/ble_nus_peripheral.rs:62]: According to the BLE spec (https://www.bluetooth.com/wp-content/uploads/Files/Specification/HTML/Core-54/out/en/host/generic-access-profile.html#UUID-8942f34d-80a7-1f7f-8f92-c8ee39aed116_informaltable-idm13358906307630), they are mandatory ("The GATT Server shall contain the GAP service as defined in the GAP Service Requirements in Table 12.1. A device shall have only one instance of the GAP service in the GATT Server. The GAP service is a GATT based service with the service UUID as «GAP Service» defined in [Assigned Numbers](https://www.bluetooth.com/specifications/assigned-numbers).")

So I don't think it should be optional.

Comment (petekubiak) [reply examples/apps/src/ble_nus_peripheral.rs:62]: Aha, thanks for the info! Turns out I hadn't properly remembered what @jamessizeland had told me - he had pointed out that devices could present as various different types of device depending on the way in which the GAP service was set up. Long story short, this is definitely functionality we'd like to add to the macro yes :)

Comment (jamessizeland) [reply examples/apps/src/ble_nus_peripheral.rs:62]: Agree we should add it either to the macro or as a builder. I'll sketch out some options tomorrow to see what's most ergonomic.

I'll open an issue to discuss.

Comment (jamessizeland) [reply examples/apps/src/ble_nus_peripheral.rs:62]: https://github.com/embassy-rs/trouble/issues/142

Comment (jamessizeland) [reply examples/apps/src/ble_nus_peripheral.rs:62]: This is completed and merged now, should no longer be blocking this for you @plaes :)

File: examples/apps/src/ble_nus_peripheral.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                Advertisement::ConnectableScannableUndirected {
+                    adv_data: &adv_data[..],
+                    scan_data: &[],
+                },
+            )
+            .await?;
+        let conn = advertiser.accept().await?;
+
+        /* TODO: Implement "echo" and push rx bytes back to tx? */
+        // XXX: I have to manually "fix" the size from 123 -> 128
```
Comment (plaes) [inline examples/apps/src/ble_nus_peripheral.rs:117]: @lulf Any clues what I'm doing wrong here? :)

**Comment by jamessizeland** [comment]
what hardware do I need to try this out?

File: examples/apps/src/ble_nus_peripheral.rs (hunk 2 above)
Comment (petekubiak) [reply examples/apps/src/ble_nus_peripheral.rs:117]: @plaes Can you explain a little more about the issue? Does something break if you make the `tx` array 123 bytes?

Comment (plaes) [reply examples/apps/src/ble_nus_peripheral.rs:117]: If you have following on line 118, it asserts that 123 != 128 when sending the notification after someone has connected:
```rust
let mut tx = [0; ATT_MTU];
```

Comment (petekubiak) [reply examples/apps/src/ble_nus_peripheral.rs:117]: Interesting...I'm guessing this is the asset in question: https://github.com/embassy-rs/trouble/blob/51d2b95a29b258b7b48f877cdf7af8a0ba54cd75/host/src/attribute.rs#L381

The issue you're facing is because your `tx` characteristic is of type `heapless::Vec<u8, 123>`, while the local variable `tx` is of type `[u8; 123]`. `heapless::Vec` has 4 bytes of metadata associated with it, giving 127 bytes which is probably being padded to 128 for alignment :)

Comment (petekubiak) [reply examples/apps/src/ble_nus_peripheral.rs:117]: Following up on this, I'd definitely recommend using the same type throughout. With your code as it is currently written, line 124 is actually modifying the first byte in the `heapless::Vec`'s length field (which is of type `usize`). This will cause undefined behaviour if you were to try to use it as a `Vec` elsewhere.

On the whole I would be reluctant to use non-primitive types for characteristics as it assumes that the other end of the comms channel has an understanding of that type. Furthermore, the heapless implementation of Vec could change its memory structure in a future version, which would change the representation of the data in memory and cause further confusion.

TL;DR: in this instance I'd strongly recommend making your `tx` characteristic a byte array.

Comment (plaes) [reply examples/apps/src/ble_nus_peripheral.rs:117]: Firstly, nice catch regarding the Vec vs byte array mismatch...

Though, using `heapless::Vec` hasn't been issue so far with nrf-softdevice library - https://github.com/lulf/watchful/blob/main/firmware/app/src/ble.rs

Comment (petekubiak) [reply examples/apps/src/ble_nus_peripheral.rs:117]: I'm not particularly familiar with nrf-softdevice, but I believe that nrf-softdevice has logic to deal with characteristics defined as `Vec` type as variable-length characteristics. So on transmission it will only send `vector.len()` bytes, and on reception it will update the vector's length to the length of the received data.
At present (correct me if I'm wrong @lulf) TrouBLE doesn't have any of this logic and treats all characteristics as a simple block of data. So your `tx` characteristic will be sent as 128 bytes regardless of the current length of the vector.

I think support for variable-length characteristics is certainly something we want to add to TrouBLE though.

Comment (plaes) [reply examples/apps/src/ble_nus_peripheral.rs:117]: This is where the assertion happens:
https://github.com/embassy-rs/trouble/blob/51d2b95a29b258b7b48f877cdf7af8a0ba54cd75/host/src/attribute.rs#L376-L383

Regarding nrf-softdevice, all the value conversions seem to go through `to_gatt()` and `from_gatt()` methods, which is provided by two traits: [`FixedGattValue`](https://github.com/embassy-rs/nrf-softdevice/blob/991d28e3c61687908acc53e748191eb59d00e27a/nrf-softdevice/src/ble/gatt_traits.rs#L12-L22) and [`GattValue`](https://github.com/embassy-rs/nrf-softdevice/blob/991d28e3c61687908acc53e748191eb59d00e27a/nrf-softdevice/src/ble/gatt_traits.rs#L24-L35).

GattValue can handle data with length smaller than attribute size and also implements support for Vec and String types via Heapless.

Comment (jamessizeland) [reply examples/apps/src/ble_nus_peripheral.rs:117]: Nice, I suspect we'll want to just implement the same, or possibly do so as part of the bt-hci crate.

Comment (petekubiak) [reply examples/apps/src/ble_nus_peripheral.rs:117]: Yes, I'm in the process of bringing the FixedGattValue and GattValue traits into TrouBLE as part of my work on #136. The purpose of that work isn't to implement support for variable length characteristics, but it will provide the framework to be able to implement this in future.

Comment (petekubiak) [reply examples/apps/src/ble_nus_peripheral.rs:117]: I've raised #149 to track the variable length characteristic implementation

Comment (lulf) [reply examples/apps/src/ble_nus_peripheral.rs:117]: Agreed bringing over FixedGattValue and GattValue traits to trouble makes sense. There is also #133 which might also allow doing varlength without having to store it.

**Comment by jamessizeland** [comment]
are you blocked by anything on finishing this PR that we can help with @plaes?

**Comment by jamessizeland** [comment]
Cool, no worries, I'll see about taking a look!

**Comment by lulf** [comment]
Is this still being worked on, or should it be closed or does anyone want to pick it up?

**Comment by plaes** [comment]
> Is this still being worked on, or should it be closed or does anyone want to pick it up?

I won't have time to deal with this until June.
